File: .opencode/skills/renpy-extract/split_translations.py
import re
import os
from pathlib import Path
from collections import defaultdict

ARCHIVE_MAPPING = {
    "Prologue": ["start", "OpeningScene"],
    "WarriorPath": ["Warrior", "Hyral", "GallowCreek"],
    "MagePath": ["Mage", "TheBlackMonolith", "TheHollowWorld"],
    "MarionPath": ["Marion", "Varga", "WarCouncil", "CouncilMeeting", "WarStrategy"],
    "SailorPath": ["Sailor", "Belmont", "CallOut", "EscapeLethram", "NoMoney"],
    "HassarPath": ["Hassar", "Jaeid", "Desert"],
    "UnionKingdom": ["Union"],
    "Relationships": ["Kate", "Lily", "Samayra", "Nick", "Carolyn", "Cassius"],
    "Bagrad": ["Bagrad", "Bribe"],
    "SakarPath": ["Sakar"],
    "DemonArc": ["DemonArc", "Razaphel"],
    "Other": []
}

def is_hex(s):
    return len(s) >= 6 and len(s) <= 8 and all(c in '0123456789abcdef' for c in s.lower())

def get_scene_name(key: str) -> str:
    parts = key.rsplit("_", 1)
    if len(parts) == 1:
        return key
    last = parts[-1]
    if last.isdigit():
        parts2 = parts[0].rsplit("_", 1)
        if len(parts2) > 1 and is_hex(parts2[-1]):
            return parts2[0]
        return parts[0]
    if is_hex(last):
        return parts[0]
    return key

def get_archive(scene_name: str) -> str:
    for archive, prefixes in ARCHIVE_MAPPING.items():
        if archive == "Other":
            continue
        for prefix in prefixes:
            if scene_name.startswith(prefix):
                return archive
    return "Other"

def parse_rpy(filepath: Path) -> dict[str, list[str]]:
    blocks = defaultdict(list)
    current_block = []
    current_key = None
    skip_next_empty = False

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            match = re.match(r'^# script\.rpy:(\d+)$', line.strip())
            if match:
                continue

            match = re.match(r'^translate ru ([A-Za-z0-9]+)_([a-f0-9]+)(?:_(\d+))?:$', line.strip())
            if match:
                if current_key:
                    blocks[current_key] = current_block
                scene_name = match.group(1)
                hash_part = match.group(2)
                suffix = match.group(3)
                current_key = f"{scene_name}_{hash_part}" if not suffix else f"{scene_name}_{hash_part}_{suffix}"
                current_block = [line]
                skip_next_empty = True
            elif current_key:
                if skip_next_empty and not line.strip():
                    skip_next_empty = False
                    continue
                current_block.append(line)
                skip_next_empty = False
            else:
                if not line.strip():
                    current_block.append(line)

    if current_key and current_block:
        blocks[current_key] = current_block

    return blocks

def split_translations(source_path: str, output_dir: str, manifest_dir: str = None) -> dict:
    source = Path(source_path)
    output = Path(output_dir)
    script_dir = Path(__file__).parent if not manifest_dir else Path(manifest_dir)
    manifest_path = script_dir / "manifest.json"

    if output.exists():
        for item in output.rglob("*"):
            if item.is_file():
                item.unlink()
        for item in list(output.rglob("*")):
            if item.is_dir() and not any(item.iterdir()):
                item.rmdir()
    output.mkdir(parents=True, exist_ok=True)

    blocks = parse_rpy(source)

    archive_files = defaultdict(list)
    total_lines = 0

    for key, lines in blocks.items():
        scene_name = get_scene_name(key)
        archive = get_archive(scene_name)
        archive_dir = output / archive
        archive_dir.mkdir(exist_ok=True)

        filepath = archive_dir / f"{scene_name}.rpy"
        with open(filepath, "a", encoding="utf-8") as f:
            f.writelines(lines)

        if scene_name not in archive_files[archive]:
            archive_files[archive].append(scene_name)
        total_lines += len(lines)

    manifest = {
        "archives": {},
        "total_scenes": len(blocks),
        "total_lines": total_lines,
        "source_lines": sum(1 for _ in open(source, encoding="utf-8"))
    }

    for archive, scenes in archive_files.items():
        manifest["archives"][archive] = {
            "count": len(scenes),
            "scenes": sorted(scenes)
        }

    manifest_path = script_dir / "manifest.json"
    import json
    with open(manifest_path, "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2, ensure_ascii=False)

    return manifest

def verify_consistency(source_path: str, output_dir: str, manifest_dir: str = None) -> dict:
    source = Path(source_path)
    output = Path(output_dir)
    script_dir = Path(__file__).parent if not manifest_dir else Path(manifest_dir)

    manifest_path = script_dir / "manifest.json"
    if not manifest_path.exists():
        return {"valid": False, "error": "manifest.json not found"}

    import json
    with open(manifest_path, encoding="utf-8") as f:
        manifest = json.load(f)

    total_output_lines = 0
    for rpy_file in output.rglob("*.rpy"):
        if rpy_file.name == "manifest.json":
            continue
        with open(rpy_file, encoding="utf-8") as f:
            total_output_lines += sum(1 for _ in f)

    return {
        "valid": manifest["total_lines"] == total_output_lines,
        "source_lines": manifest["source_lines"],
        "output_lines": total_output_lines,
        "manifest_total_lines": manifest["total_lines"],
        "manifest_source_lines": manifest["source_lines"]
    }

if __name__ == "__main__":
    import argparse
    from pathlib import Path

    script_dir = Path(__file__).parent
    game_dir = script_dir.parent.parent.parent

    parser = argparse.ArgumentParser(description="Split/verify Ren'Py translations")
    parser.add_argument("--source", default=str(game_dir / "game" / "tl" / "ru" / "script" / "__.script.__rpy"),
                        help="Source translation file")
    parser.add_argument("--output", default=str(game_dir / "game" / "tl" / "ru" / "script" / "split"),
                        help="Output directory for split files")
    parser.add_argument("--manifest", default=str(script_dir / "manifest.json"),
                        help="Path to manifest.json")
    parser.add_argument("command", nargs="?", default="split", choices=["split", "verify"],
                        help="Command: split or verify")

    args = parser.parse_args()

    if args.command == "verify":
        result = verify_consistency(args.source, args.output)
        print(f"Valid: {result['valid']}")
        print(f"Source lines: {result['source_lines']}")
        print(f"Output lines: {result['output_lines']}")
    else:
        manifest = split_translations(args.source, args.output)
        print(f"Split {manifest['total_scenes']} scenes into {args.output}")
        for archive, data in manifest["archives"].items():
            print(f"  {archive}: {data['count']} scenes")

File: .opencode/skills/renpy-extract/split_translations.mjs
#!/usr/bin/env node

/**
 * @module split_translations
 * @description Разделение/верификация Ren'Py переводов.
 * Разбивает единый файл переводов на файлы по аркам/сценам.
 *
 * Требования:
 * - Node.js >= 18 (использует fs.readFileSync, fs.writeFileSync, fs.readdirSync,
 *   fs.existsSync, fs.mkdirSync, path.join, path.relative)
 * - Никаких внешних зависимостей
 *
 * Использование:
 *   node split_translations.mjs split
 *   node split_translations.mjs verify
 *   node split_translations.mjs --source <path> --output <dir> split
 *
 * @see {@link https://nodejs.org/docs/latest-v18.x/api/fs.html|Node.js fs docs}
 */

import { readFileSync, writeFileSync, readdirSync, existsSync, mkdirSync, rmSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * @type {Object<string, string[]>}
 */
const ARCHIVE_MAPPING = {
  Prologue: ['start', 'OpeningScene'],
  WarriorPath: ['Warrior', 'Hyral', 'GallowCreek'],
  MagePath: ['Mage', 'TheBlackMonolith', 'TheHollowWorld'],
  MarionPath: ['Marion', 'Varga', 'WarCouncil', 'CouncilMeeting', 'WarStrategy'],
  SailorPath: ['Sailor', 'Belmont', 'CallOut', 'EscapeLethram', 'NoMoney'],
  HassarPath: ['Hassar', 'Jaeid', 'Desert'],
  UnionKingdom: ['Union'],
  Relationships: ['Kate', 'Lily', 'Samayra', 'Nick', 'Carolyn', 'Cassius'],
  Bagrad: ['Bagrad', 'Bribe'],
  SakarPath: ['Sakar'],
  DemonArc: ['DemonArc', 'Razaphel'],
  Other: [],
};

/**
 * Разбивает строку по последнему разделителю (аналог Python str.rsplit).
 * @param {string} str
 * @param {string} separator
 * @param {number} [limit]
 * @returns {string[]}
 */
function rsplit(str, separator, limit) {
  const arr = str.split(separator);
  if (limit === undefined || arr.length <= limit) return arr;
  const last = arr.slice(-limit);
  const first = arr.slice(0, arr.length - limit).join(separator);
  return [first, ...last];
}

/**
 * Проверяет, является ли строка hex (6-8 символов).
 * @param {string} s
 * @returns {boolean}
 */
function isHex(s) {
  return s.length >= 6 && s.length <= 8 && /^[0-9a-f]+$/i.test(s);
}

/**
 * Извлекает имя сцены из ключа блока.
 * @param {string} key - например "HyralOrc2_c3fc8560" или "start_2b88e3eb"
 * @returns {string}
 */
export function getSceneName(key) {
  const parts = rsplit(key, '_', 1);
  if (parts.length === 1) return key;

  const last = parts[parts.length - 1];
  if (/^\d+$/.test(last)) {
    const parts2 = rsplit(parts[0], '_', 1);
    if (parts2.length > 1 && isHex(parts2[parts2.length - 1])) {
      return parts2[0];
    }
    return parts[0];
  }
  if (isHex(last)) {
    return parts[0];
  }
  return key;
}

/**
 * Определяет архив (арку) по имени сцены.
 * @param {string} sceneName
 * @returns {string}
 */
export function getArchive(sceneName) {
  for (const [archive, prefixes] of Object.entries(ARCHIVE_MAPPING)) {
    if (archive === 'Other') continue;
    for (const prefix of prefixes) {
      if (sceneName.startsWith(prefix)) return archive;
    }
  }
  return 'Other';
}

/**
 * Парсит .rpy файл на блоки translate ru <scene>_<hash>:
 * @param {string} filepath
 * @returns {Object<string, string[]>}
 */
export function parseRpy(filepath) {
  /** @type {Object<string, string[]>} */
  const blocks = {};
  let currentBlock = null;
  let currentKey = null;
  let skipNextEmpty = false;

  const content = readFileSync(filepath, 'utf-8');
  const lines = content.split('\n');

  for (const line of lines) {
    // Пропускаем # script.rpy:N строки
    if (/^# script\.rpy:\d+$/.test(line.trim())) continue;

    const match = /^translate ru ([A-Za-z0-9]+)_([a-f0-9]+)(?:_(\d+))?:$/.exec(line.trim());
    if (match) {
      if (currentKey && currentBlock) {
        blocks[currentKey] = currentBlock;
      }
      const sceneName = match[1];
      const hashPart = match[2];
      const suffix = match[3];
      currentKey = suffix !== undefined ? `${sceneName}_${hashPart}_${suffix}` : `${sceneName}_${hashPart}`;
      currentBlock = [line];
      skipNextEmpty = true;
    } else if (currentKey) {
      if (skipNextEmpty && line.trim() === '') {
        skipNextEmpty = false;
        continue;
      }
      currentBlock.push(line);
      skipNextEmpty = false;
    } else {
      if (line.trim() === '') {
        // Keep empty lines outside blocks
      }
    }
  }

  if (currentKey && currentBlock) {
    blocks[currentKey] = currentBlock;
  }

  return blocks;
}

/**
 * Разделяет исходный файл переводов на файлы по аркам.
 * @param {string} sourcePath - путь к исходному файлу
 * @param {string} outputDir - выходная директория
 * @param {string} [manifestDir] - директория для manifest.json
 * @returns {object} manifest
 */
export function splitTranslations(sourcePath, outputDir, manifestDir) {
  const source = sourcePath;
  const output = outputDir;
  const scriptDir = manifestDir || dirname(fileURLToPath(new URL('.', import.meta.url)));
  const manifestPath = join(scriptDir, 'manifest.json');

  // Clean output dir
  if (existsSync(output)) {
    cleanDir(output);
  }
  mkdirSync(output, { recursive: true });

  const blocks = parseRpy(source);

  /** @type {Object<string, string[]>} */
  const archiveFiles = {};
  let totalLines = 0;

  for (const [key, lines] of Object.entries(blocks)) {
    const sceneName = getSceneName(key);
    const archive = getArchive(sceneName);
    const archiveDir = join(output, archive);
    mkdirSync(archiveDir, { recursive: true });

    const filepath = join(archiveDir, `${sceneName}.rpy`);
    writeFileSync(filepath, lines.join('\n') + '\n', { flag: 'a', encoding: 'utf-8' });

    if (!archiveFiles[archive]) archiveFiles[archive] = [];
    if (!archiveFiles[archive].includes(sceneName)) {
      archiveFiles[archive].push(sceneName);
    }
    totalLines += lines.length;
  }

  const sourceContent = readFileSync(source, 'utf-8');
  const manifest = {
    archives: {},
    total_scenes: Object.keys(blocks).length,
    total_lines: totalLines,
    source_lines: sourceContent.split('\n').length,
  };

  for (const [archive, scenes] of Object.entries(archiveFiles)) {
    manifest.archives[archive] = {
      count: scenes.length,
      scenes: scenes.sort(),
    };
  }

  writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), 'utf-8');

  return manifest;
}

/**
 * Рекурсивно очищает директорию.
 * @param {string} dir
 */
function cleanDir(dir) {
  if (!existsSync(dir)) return;
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      cleanDir(fullPath);
      try { rmSync(fullPath); } catch { /* ignore */ }
    } else {
      try { rmSync(fullPath); } catch { /* ignore */ }
    }
  }
}

/**
 * Проверяет целостность сплита.
 * @param {string} sourcePath
 * @param {string} outputDir
 * @param {string} [manifestDir]
 * @returns {object}
 */
export function verifyConsistency(sourcePath, outputDir, manifestDir) {
  const source = sourcePath;
  const output = outputDir;
  const scriptDir = manifestDir || dirname(fileURLToPath(new URL('.', import.meta.url)));
  const manifestPath = join(scriptDir, 'manifest.json');

  if (!existsSync(manifestPath)) {
    return { valid: false, error: 'manifest.json not found' };
  }

  const manifest = JSON.parse(readFileSync(manifestPath, 'utf-8'));

  let totalOutputLines = 0;

  function countLines(dir) {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        countLines(fullPath);
      } else if (entry.isFile() && entry.name.endsWith('.rpy')) {
        const content = readFileSync(fullPath, 'utf-8');
        const lines = content.split('\n');
        const count = content.endsWith('\n') ? lines.length - 1 : lines.length;
        totalOutputLines += count;
      }
    }
  }
  countLines(output);

  return {
    valid: manifest.total_lines === totalOutputLines,
    source_lines: manifest.source_lines,
    output_lines: totalOutputLines,
    manifest_total_lines: manifest.total_lines,
    manifest_source_lines: manifest.source_lines,
  };
}

/**
 * CLI entry point.
 */
function main() {
  const args = process.argv.slice(2);

  const scriptDir = dirname(fileURLToPath(new URL('.', import.meta.url)));
  const gameDir = join(scriptDir, '..', '..', '..');

  let sourcePath = join(gameDir, 'game', 'tl', 'ru', 'script', '__.script.__rpy');
  let outputDir = join(gameDir, 'game', 'tl', 'ru', 'script', 'split');
  let manifestDir = scriptDir;
  let command = 'split';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--source') sourcePath = args[++i];
    else if (args[i] === '--output') outputDir = args[++i];
    else if (args[i] === '--manifest') manifestDir = args[++i];
    else if (args[i] === 'split' || args[i] === 'verify') command = args[i];
  }

  if (command === 'verify') {
    const result = verifyConsistency(sourcePath, outputDir, manifestDir);
    console.log(`Valid: ${result.valid}`);
    console.log(`Source lines: ${result.source_lines}`);
    console.log(`Output lines: ${result.output_lines}`);
  } else {
    const manifest = splitTranslations(sourcePath, outputDir, manifestDir);
    console.log(`Split ${manifest.total_scenes} scenes into ${outputDir}`);
    for (const [archive, data] of Object.entries(manifest.archives)) {
      console.log(`  ${archive}: ${data.count} scenes`);
    }
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}
